base/broadcast.jl

```diff
@@ -272,19 +297,28 @@ end
 ftype(f, A) = typeof(f)
 ftype(f, A...) = typeof(a -> f(a...))
 ftype(T::Type, A...) = Type{T}
-typestuple(a) = (Base.@_pure_meta; Tuple{eltype(a)})
-typestuple(T::Type) = (Base.@_pure_meta; Tuple{Type{T}})
-typestuple(a, b...) = (Base.@_pure_meta; Tuple{typestuple(a).types..., typestuple(b...).types...})
-ziptype(A) = typestuple(A)
-ziptype(A, B) = (Base.@_pure_meta; Iterators.Zip2{typestuple(A), typestuple(B)})
-@inline ziptype(A, B, C, D...) = Iterators.Zip{typestuple(A), ziptype(B, C, D...)}
 
-_broadcast_type(f, T::Type, As...) = Base._return_type(f, typestuple(T, As...))
-_broadcast_type(f, A, Bs...) = Base._default_eltype(Base.Generator{ziptype(A, Bs...), ftype(f, A, Bs...)})
+# nullables need to be treated like scalars sometimes and like containers
+# other times, so there are two variants of typestuple.
+immutable NullableIsScalar end
+immutable NullableIsContainer end
+
+typestuple(::Any, a) = (Base.@_pure_meta; Tuple{eltype(a)})
+typestuple(::NullableIsScalar, a::Nullable) = (Base.@_pure_meta; Tuple{typeof(a)})
```

Comment:
I like this solution much better :+1:, but I don't think you need to introduce a new type here, you could just use `Array` and `Any`. That is, `typestuple(::Type{Array}, a::Nullable) = (Base.@_pure_meta; Tuple{typeof(a)})` and `T = _broadcast_type(Array, f, A, Bs...)` or `T = _broadcast_type(Any, f, A, Bs...)`.